vm/actor/src/builtin/miner/miner_state.rs

```diff
@@ -0,0 +1,993 @@
+        expire_epoch: ChainEpoch,
+        sector_number: SectorNumber,
+    ) -> Result<(), Box<dyn StdError>> {
+        // Load BitField Queue for sector expiry
+        let quant = self.quant_spec_every_deadline();
+        let mut queue = super::BitFieldQueue::new(store, &self.pre_committed_sectors_expiry, quant)
+            .map_err(|e| format!("failed to load pre-commit sector queue: {:?}", e))?;
+
+        // add entry for this sector to the queue
+        queue.add_to_queue_values(expire_epoch, &[sector_number])?;
+        self.pre_committed_sectors_expiry = queue.amt.flush()?;
+
+        Ok(())
+    }
+
+    pub fn check_precommit_expiry<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        sectors: &BitField,
+    ) -> Result<TokenAmount, HamtError> {
+        let mut deposit_to_burn = TokenAmount::zero();
+        let mut precommits_to_delete = Vec::new();
+
+        for i in sectors.iter() {
+            let sector_number = i as SectorNumber;
+
+            let sector = match self.get_precommitted_sector(store, sector_number)? {
+                Some(sector) => sector,
+                // already committed/deleted
+                None => continue,
+            };
+
+            // mark it for deletion
+            precommits_to_delete.push(sector_number);
+
+            // increment deposit to burn
+            deposit_to_burn += sector.pre_commit_deposit;
+        }
+
+        // Actually delete it.
+        if !precommits_to_delete.is_empty() {
+            self.delete_precommitted_sector(store, &precommits_to_delete)?;
+        }
+
+        self.pre_commit_deposits -= &deposit_to_burn;
+        assert!(self.pre_commit_deposits >= TokenAmount::zero());
+
+        Ok(deposit_to_burn)
+    }
+}
+
+/// Static information about miner
+#[derive(Debug, PartialEq, Serialize_tuple, Deserialize_tuple)]
+pub struct MinerInfo {
+    /// Account that owns this miner
+    /// - Income and returned collateral are paid to this address
+    /// - This address is also allowed to change the worker address for the miner
+    pub owner: Address,
+
+    /// Worker account for this miner
+    /// This will be the key that is used to sign blocks created by this miner, and
+    /// sign messages sent on behalf of this miner to commit sectors, submit PoSts, and
+    /// other day to day miner activities
+    pub worker: Address,
+
+    /// Additional addresses that are permitted to submit messages controlling this actor (optional).
+    pub control_addresses: Vec<Address>, // Must all be ID addresses.
+
+    /// Optional worker key to update at an epoch
+    pub pending_worker_key: Option<WorkerKeyChange>,
+
+    /// Libp2p identity that should be used when connecting to this miner
+    #[serde(with = "serde_bytes")]
+    pub peer_id: Vec<u8>,
+    /// Slice of byte arrays representing Libp2p multi-addresses used for establishing a connection with this miner.
+    #[serde(with = "serde_bytes")]
+    pub multi_address: Vec<u8>,
```

Comment:
This should be a 2d vector, a MultiAddress is a byte slice, and it's a vector of them